contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,572 @@
+Persistent Volume API.
+
+WORKFLOW:
+
+USER:
+
+```
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: local-raw-pvc
+spec:
+  volumeMode: Block
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 80Gi
+```
+
+## UC4: 
+
+DESCRIPTION: StorageClass with non-dynamically created volumes
+
+BACKGROUND: The admin wishes to create a storage class that will identify pre-provisioned block PVs based on a user's PVC request for volumeMode: Block. 
+
+WORKFLOW: 
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: block-volume
+provisioner: no-provisioning 
+parameters:
+```
+* Sample of pre-created volume definition:
+
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+ name: pv-block-volume
+spec:
+ volumeMode: Block
+ storageClassName: block-volume
+ capacity:
+   storage: 35Gi
+ accessModes:
+   - ReadWriteOnce
+ local:
+    path: /dev/xvdc
+```
+## [FUTURE] UC5: 
+
+DESCRIPTION: StorageClass with dynamically created volumes 
+
+BACKGROUND: The admin wishes to create a storage class that will dynamically create block PVs based on a user's PVC request for volumeMode: Block. The admin desires the volumes be created dynamically and deleted when the PV definition is deleted. 
+
+WORKFLOW:
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: local-fast
+provisioner: kubernetes.io/local-block-ssd
+parameters:
+  fstype: Block #suggested value
+```
+
+***This has implementation details that have yet to be determined. It is included in this proposal for completeness of design ****
+
+## UC6: 
+
+DESCRIPTION: The developer wishes to request a block device via a Storage Class.
+
+WORKFLOW:
+
+USER: 
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+ name: pvc-local-block
+spec:
+ volumeMode: Block
+ storageClassName: local-fast
+ accessModes:
+  - ReadWriteOnce
+ resources:
+   requests:
+     storage: 10Gi
+```  
+
+## UC7: 
+
+DESCRIPTION: Admin creates network raw block devices
+
+BACKGROUND: Admin wishes to pre-create Persistent Volumes in GCE as raw block devices
+
+WORKFLOW:
+
+ADMIN:
+
+```
+apiVersion: "v1"
+kind: "PersistentVolume"
+metadata:
+  name: gce-disk-1
+  annotations:
+    volume.beta.kubernetes.io/mount-options: "discard"
+Spec:
+  volumeMode: Block
+  capacity:
+    storage: "10Gi"
+  accessModes:
+    - "ReadWriteOnce"
+  gcePersistentDisk:
+    fsType: "Block" #this is a suggestion, it is plugin dependent
+    pdName: "gce-disk-1"
+```
+***fsType values will be provisioner dependent. Block is suggested for development simplicity. Since the PVC object is passed
+   to the provisioner, it will be responsible for validating and handling whether or not it supports the volumeMode being passed ***
+
+## UC8: 
+
+DESCRIPTION: 
+
+A developer wishes to enable their application to use a raw block device as an inline volume in the pod. 
```

Comment:
@saad-ali mentioned that you were not going to support inline volumes after all, and only PVC/PV.

Reply:
Only adding these changes to PV/PVC would probably remove the need for PSP changes… the existing controls around PVC use and quota around use of specific storage classes would probably be sufficient

Reply:
@msau42 well, I mentioned this to @saad-ali. There is a limitation of inline volumes in that it would equate the device path and container path to the same path in the CRI. I don't really care for this, but it doesn't make it a show stopper either. I honestly do not think it's a requirement that we do and if local doesn't support it either, it's an even stronger case. If you agree, I will just remove this.

Reply:
@liggitt that is an excellent point and really should be the first line of control for storage.

Reply:
Why would the device path and container path need to be the same?   The container path would just be the volumeDevices.containerPath, and the device path would be plugin dependent.  At the plugin level, there shouldn't be a difference between inline volumes and PVs in terms of the device path.

Anyway, I'm totally fine with not allowing it for inline volumes.  Most of our new storage features don't support inline volumes anyway, and it seems like this will eliminate the need for PSP changes.